src/disk_io_thread.cpp

```diff
@@ -193,63 +197,25 @@ namespace libtorrent
 	void disk_io_thread::abort(bool wait)
 	{
 		m_abort = true;
-		if (m_num_threads == 0)
+		if (num_threads() == 0)
 		{
 			abort_jobs();
 		}
 		else
 		{
-			set_num_threads(0, wait);
+			m_generic_threads.abort(wait);
+			m_hash_threads.abort(wait);
 		}
 	}
 
 	// TODO: 1 it would be nice to have the number of threads be set dynamically
-	void disk_io_thread::set_num_threads(int const i, bool const wait)
+	void disk_io_thread::set_num_threads(int const i)
 	{
 		TORRENT_ASSERT(m_magic == 0x1337);
-		if (i == m_num_threads) return;
 
-		if (i > m_num_threads)
-		{
-			while (m_num_threads < i)
-			{
-				int thread_id = (++m_num_threads) - 1;
-				thread_type_t type = generic_thread;
-
-				// this keeps the io_service::run() call blocked from returning.
-				// When shutting down, it's possible that the event queue is drained
-				// before the disk_io_thread has posted its last callback. When this
-				// happens, the io_service will have a pending callback from the
-				// disk_io_thread, but the event loop is not running. this means
-				// that the event is destructed after the disk_io_thread. If the
-				// event refers to a disk buffer it will try to free it, but the
-				// buffer pool won't exist anymore, and crash. This prevents that.
-				boost::shared_ptr<io_service::work> work =
-					boost::make_shared<io_service::work>(boost::ref(m_ios));
-
-				// the magic number 3 is also used in add_job()
-				// every 4:th thread is a hasher thread
-				if ((thread_id & 0x3) == 3) type = hasher_thread;
-				m_threads.emplace_back(&disk_io_thread::thread_fun, this
-						, thread_id, type, work);
-			}
-		}
-		else
-		{
-			while (m_num_threads > i) { --m_num_threads; }
-			std::unique_lock<std::mutex> l(m_job_mutex);
-			m_job_cond.notify_all();
-			m_hash_job_cond.notify_all();
-			l.unlock();
-			for (int j = m_num_threads; j < m_threads.size(); ++j)
-			{
-				if (wait)
-					m_threads[j].join();
-				else
-					m_threads[j].detach();
-			}
-			m_threads.resize(m_num_threads);
-		}
+		// add one hasher thread for every three generic threads
+		m_generic_threads.set_max_threads(i - i / 4);
```

Comment:
I was fooled by this expression, thinking "i - i", that's zero! I think it would be nice to say:

int const num_hash_threads = i / 4;
int const num_generic_threads = i - num_hash_threads;
...